cody.el

```diff
@@ -1179,49 +1327,52 @@ handles logging and messaging."
                   (let ((output (apply #'message msg args)))
                     (if verbose (message output)
                       (cody--log output))
-                    'fail))) ; return non-nil to signal failed check
+                    nil))) ; return nil to signal failed check
       (cond
-       ((not cody-enable-completion-cycling)
-        (explain "Set `cody-enable-completion-cycling' to enable cycling."))
-       ((not (and (cody--cc) (cody--current-item (cody--cc))))
+       ((not cody-completions-enable-cycling-p)
+        (explain "Set `cody-completions-enable-cycling-p' to enable cycling."))
+       ((not (and cc (cody--current-item cc)))
         (explain "No completion here"))
        ((not (cody--multiple-items-p cc))
         (explain "No other suggestions; %s"
                  (cody--key-msg-for-command 'cody--tab-key)))
-       ;; Disallow cycling if they have typed into the completion prefix.
-       ((not (string= (cody--completion-display-text cc)
-                      (cody--completion-original-text cc)))
-        (explain "Cycling currently unavailable."))
-       (t nil))))) ; cycling is available
+       (t 'success))))) ; cycling is not available
 
 (defun cody--cycle-completion (direction)
   "Cycle through the completion items in the specified DIRECTION.
 DIRECTION should be 1 for next, and -1 for previous."
-  (unless (cody--check-cycle-preconditions)
-    (if-let* ((cc (cody--cc))
-              (items (cody--completion-items cc))
-              (index (cody--current-item-index cc))
-              (num (cody--num-items cc))
-              (next (% (+ index direction num) num)))
-        (progn
-          (setf (cody--current-item-index cc) next)
-          (cody--display-completion next))
-      (when cody-enable-completion-cycling-help
-        (message "Error cycling through completions"))
-      (cody--log "Error cycling through completions: items= %s" items))))
+  (if (cody--check-cycle-preconditions)
+      (if-let* (
+                (cc (cody--cc))
+                (items (cody--completion-items cc))
+                (index (cody--current-item-index cc))
+                (num (cody--num-items cc))
+                (next (% (+ index direction num) num)))
+          (condition-case err
+              (progn
+                (cody--hide-completion) ; This sets index to 0.
+                (setf (cody--current-item-index cc) next) ; Set index to new val.
+                (cody--display-completion))
+            (error
+             (cody--log "Error displaying completion: %s \ncompletion: %s" err cc)))
+        (when cody-completions-cycling-help-p
+          (message "Error cycling through completions"))
+        (cody--log "Error cycling through completions: items= %s" items))
+    (when cody-completions-cycling-help-p
+      (message "No more suggestions. TAB to accept."))))
```

Comment:
Given the infinite rebindability of stuff, is it really always TAB?

Reply:
I've addressed key rebindings in a follow-up commit.